torch/fx/subgraph_rewriter.py

```diff
@@ -0,0 +1,202 @@
+    # been used as part of another match
+    def overlaps_with_prev_match(match: Match) -> bool:
+        for n in match.nodes_map.values():
+            if n in nodes_to_delete:
+                return True
+        return False
+
+    if not matches:
+        raise ValueError("No matches were found for the given pattern")
+
+    for match in matches:
+
+        # Skip overlapping matches
+        if overlaps_with_prev_match(match):
+            continue
+
+        # Get the original graph nodes corresponding to the "start" and
+        # "end" of the pattern subgraph. These connections are where
+        # we'll hook the replacement subgraph in to the original graph
+        original_graph_inputs: Dict[str, Node] = {n.name : n
+                                                  for n
+                                                  in original_graph.nodes
+                                                  if n.op == "placeholder"}
+        subgraph_output: Node = match.anchor
+
+        def mark_nodes_for_deletion(n: Node) -> None:
+            if n not in match.nodes_map.values():
+                return
+            for n_ in n.uses:
+                mark_nodes_for_deletion(n_)
+            nodes_to_delete.add(n)
+            # By removing the `original_graph` node's name early, we
+            # ensure that we can insert nodes from `replacement` with
+            # the same name. (This is strictly cosmetic)
+            if n.name in original_graph._used_names:
+                original_graph._used_names.pop(n.name)
+
+        mark_nodes_for_deletion(subgraph_output)
+
+        # Map replacement graph nodes to their copy in `original_graph`
+        copied_nodes: Dict[Node, Node] = {}
+
+        # Insert the new graph
+        with original_graph.inserting_before(subgraph_output):
+            for n in replacement_graph.nodes:
+                if n.op == "output":
+                    # To ensure that we have one and only one "output"
+                    # node in our final graph, we need to change our
+                    # logic slightly if the pattern extends to the end
+                    # of the graph (i.e. if we matched the pattern
+                    # "output" node to the original graph's "output"
+                    # node)
+                    if subgraph_output.op != "output":
+                        n = n.args[0]
+                        nodes_to_delete.add(copied_nodes[n])
+                    copied_nodes[n] = original_graph.node_copy(n,
+                                                               lambda n :
+                                                               copied_nodes[n])
+                    subgraph_output.replace_all_uses_with(copied_nodes[n])
+                elif n.op == "placeholder":
+                    # Temporarily change the insertion point to the
+                    # beginning of the graph to ensure that placeholder
+                    # nodes are inserted before anything else
```

Comment:
`placeholder `in the pattern graph is essentially the base case, so the logic here should probably be to remap values in the replacement pattern with the values that correspond to the `placeholder` in the original graph that matched to a `placeholder` in the pattern graph. (Don't know if that made sense, let me know if you need a better explanation lol)